File: stat_mod.f
!gfortran, gcc version 5.4.0 20160609

!Программа статистического моделирования
program stat_mod

        !Ввод истодини данных      
        real lambda(50) , nu(50), mg(20), kg, ns, ns1, lambd, nui
        integer*2 mdi(50)
        integer*2 iy
        logical md(50,50),c(50),ch,cp,cq
        dimension t(50),tau(50),pb(50),nel(50),a(10)
        write(*,*) 'Исходные данные'
        write(*,*) '---------------'
        write(*,*) 'Введите допустимую ошибку'
        read(*,*) dd
        write(*,100) dd
100     format(1x, 'Допустимая ошибка ', f6.4)
        write(*,*) 'Введите доверительную вероятность'
        read(*,*) delta
        write(*,101) delta
101     format(1x, 'Доверительная вероятность ', f6.4)
        write(*,*) 'Введите проолжительность режима'
        read(*,*) ta
        write(*,102) ta
102     format(1x, 'Продолжительность режима', f6.0)    
        !Вычисление функции Z по данным б   
        zdelta=2569.449*delta**3-7196.125*delta**2+6722.0792*delta-2092.4842    
        write(*,*) 'Введите количество элементов системы'
        read(*,*) m
        write(*,*) 'Количество элементов системы', m
        write(*,*) 'Bведите количество условий отказа системы'
        read(*,*) kcm
        write(*,*) 'Количество условий отказа системы ', kcm
        write(*,*) 'Введите количество установленных источников'
        read(*,*) ki
        write(*,*) 'Количество установленных источников', ki
        write(*,*) 'Введите потребляемую в режиме мощность'
        read(*,*) wd
        write(*,103) wd
103     format(1x, 'Потребляемая в режиме мощность ', f7.0)
        write(*,*) 'Введите массив источников'      
        read(*,*) (mg(i),i=1,ki)
        write(*,*) 'Массив источников:'
        do 105 iii=1, ki
        write(*,106) iii, mg(iii)    
106     format(1x, 'Мощность ',i2, '-го источника ',f7.0)
105     continue
        write(*,*) 'Ввод лямбда, ну, Рв'
        write(*,*) '---------------'
1105    jn=0
1102    continue
        write(*,*) 'Введите номер элемента'
        read(*,*) neli
        if(neli.eq.0) goto 1102
        jn=jn+1
        nel(jn)=neli
        goto 1101
1101    continue
        write(*,*) 'Введите: лямбда, ну, Рв'
        read(*,*) lambd, nui,pbi
        do 1103 jj=1,jn
        kkk=nel(jj)
        lambda(kkk)=lambd
        nu(kkk)=nui
1103    pb(kkk)=pbi
        sum=sum+jn
        if(sum.lt.m) goto 1105
        write(*,*) 'Диагностическая матрица'
        write(*,*) 'Вввод Диагностической матрицы:'
        write(*,*) '---------------'    
        do 10 i=1, kcm
        write(*,110) i
110     format(1x, 'Введите номера столбцов' ,i2,'-ой строки')
        read(*,*) nn1,nn2
        do 2001 ne=1,m
2001    mdi(ne)=0
        mdi(nn1)=1
        mdi(nn2)=1
        write(*,7) (mdi(j),j=1,m)
7       format(1x,79i1)
        do 10 j=1,m
        if(mdi(j)) 8,8,9
8       md(i,j)=.false.
        goto 10
9       md(i,j)=.true.
10      continue
        write(*,*) 'При определении источников'
        write(*,*) 'проверять состояние их щитов?'
        write(*,*) 'Да - введите 1'
        write(*,*) 'Нет - введите 0'
        read(*,*) otv
        write(*,*) '-----------------------------------'
        write(*,*) '|  №  |  лямбда |    ню   |  Рв   |'
        write(*,*) '-----------------------------------'      
        do 400 j=1, m
        amb=lambda(j)*1e6
        u1=nu(j)*1e3
        write(*, 401) j, amb,u1,pb(j)
401     format(1x, '| ',i3, ' | ', f7.2, ' | ',f7.2,' | ',f5.3, ' |')
400     continue
        write(*,*) '-----------------------------------' 
        write(*,*) '     РЕЗУЛЬТАТЫ МОДЕЛИРОВАНИЯ'       
        write(*,*) '---------------------------------------'
        write(*,*) '| Д |  дельта | n исп. |  кг  |    В  |'
        write(*,*) '---------------------------------------'
  
        iy=0
        ts=0    ! Установка начальных значений
        ns=100  !времени, количества спттавки
        sqt=0.  ! и кол-ва отказов
        l=0.
12      continue
        do 32 n=1, 100 !цикл испытаний (походов судна)
            do 13 i=1, m !цикл перебора элементов системы
				c(i)=.true.
13          continue
            t(i)=0.
            th=0
            tsl=0
            cp=.true.
            cq=.true.
            do 14 i=1,m ! Определение времени
3001			continue
				call random_number(z) ! до наступления
				if(z.le.0) goto 3001 !отказов каждого
				tau(i)=-alog(z)/lambda(i) !из элементов системы.
14          continue
			t(i)=tau(i)     !определение минимального интервала времени
15          continue
			tm=t(1) ! (ts тут или нет хз) отказов элементов и номера 
            k=1 !элемента отказавшего первым
            do 16 i=2, m
				if(t(i).ge.tm) goto 16
				tm=t(i)
				k=i
16          continue
            if(tm.le.ta) goto 37
				if(cp) goto 31
				ti=ta-th  !счет времени неработоспособного состяния системы
				ts=ts+ti
				tsl=tsl+ti
				goto 31
37          continue
			if(c(k)) goto 38
				c(k)=.true.    !если у элемента к произошло восстановление, моделируется интервал времени до очередного отказа        
				call random_number(z)
				tau(k)=-alog(z)/lambda(k)
				go to 21
38          continue
			c(k)=.false. ! если у элемента произошел отказ, определяется возможность его(отказа) устранения
			call random_number(z)
            
            !write(*,*) 'd1:',z, pb(k)
            if(z.ge.pb(k)) goto 20
				call random_number(z)
                    !write(*,*) 'd2:',z, pb(k)
					tau(k)=-alog(z)/nu(k)
				goto 21
20          continue
			tau(k)=ta
21			continue          
			t(k)=t(k)+tau(k)
            if(c(k).and.cp) goto 22
				go to 23
22			continue          
			ch=.true.
            go to 28
23			continue          
			do 25 j=1, kcm
				do 24 i=1,m
					if(.not.c(i).or..not.md(j,i)) go to 24 !определение состояния системы по результату сравненения ее состояния со строками диагностической матрицы
						ch=.true.
						go to 25
24          	continue
				ch=.false.
				go to 28
25          continue
26          continue
			w=0
            do 27 i=1, ki
                if(otv) 2000,2000,3000
3000		        continue        
                    if(c(i).and.c(ki+i)) w=w+mg(i) !Суммирование мощности исправных источников
                    goto 27
2000		        continue        
                    if(c(i)) w=w+mg(i)
27          continue
            ch=.true.
            if (w.lt.wd) ch=.false.
28          continue
			if (cp.or..not.ch) goto 29 !Опеределение интервала времени неработоспособного состояния
				ti=tm-th
				ts=ts+ti
				tsl=tsl+ti
29          continue
			if(.not.cp.or.ch) goto 30
				th=tm       ! Фиксируется время неработоспособного состояния
				cq=.false.
30			continue          
			cp=ch 
            go to 15
31			continue          
			sqt=sqt+tsl*tsl 
            if(.not.cq) l=l+1   
32      continue
        kg=l-ts/ta/ns  !Вычисление коэф. готовности системы
        r=1.-l*1./ns  ! Вычисление вероятности безотказной работы 
        d=zdelta*sqrt(r*(l-r)/ns)  !значение доверительной ошибки
        if(l.gt.0.and.d.le.dd) go to 34
            write(*,33) d, delta, ns, kg, r
33          format(1x,'| ',f7.5,' | ',f5.3, ' | ' ,f6.0,' | ', f7.5,' | ', f7.5, ' |')
            ns=ns+100
            go to 12
34      continue
        write(*,33) d, delta, ns, kg, r
        write(*,*) '---------------------------------------'   
        stop
end program stat_mod

File: stat_mod.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <random>
#include <cmath>
#include <limits>
#include "parse.h"

inline std::string fortranF(double v, int width, int prec) {
    std::ostringstream ss;
    if(prec==0) ss << std::showpoint;
    ss << std::fixed << std::setprecision(prec) << v;
    std::string s = ss.str();
    if((int)s.size() > width) return std::string(width, '*');
    if((int)s.size() < width) s = std::string(width - s.size(),' ') + s;
    return s;
}

int main() {
    const int MAX_M = 50;
    const int MAX_K = 50;
    const int MAX_G = 20;

    std::vector<double> lambda(MAX_M+1, 0.0);
    std::vector<double> nu(MAX_M+1, 0.0);
    std::vector<double> mg(MAX_G+1, 0.0);
    std::vector<int> mdi(MAX_M+1, 0);
    std::vector<std::vector<bool>> md(MAX_K+1, std::vector<bool>(MAX_M+1,false));
    std::vector<bool> c(MAX_M+1, false);
    std::vector<double> t(MAX_M+1,0.0), tau(MAX_M+1,0.0), pb(MAX_M+1,0.0), nel(MAX_M+1,0.0);

    double dd, delta, ta, zdelta, wd;
    int m, kcm, ki;

    std::cout << "Исходные данные" << std::endl;
    std::cout << "---------------" << std::endl;
    std::cout << "Введите допустимую ошибку" << std::endl;
    std::cin >> dd;
    std::cout << " Допустимая ошибка " << std::fixed << std::setprecision(4) << dd << std::endl;
    std::cout << "Введите доверительную вероятность" << std::endl;
    std::cin >> delta;
    std::cout << " Доверительная вероятность " << std::fixed << std::setprecision(4) << delta << std::endl;
    std::cout << "Введите проолжительность режима" << std::endl;
    std::cin >> ta;
    std::cout << " Продолжительность режима " << std::showpoint << std::fixed << std::setprecision(0) << ta << std::endl;
    std::cout << std::noshowpoint;

    zdelta = 2569.449*std::pow(delta,3) - 7196.125*delta*delta + 6722.0792*delta - 2092.4842;

    std::cout << "Введите количество элементов системы" << std::endl;
    std::cin >> m;
    std::cout << " Количество элементов системы " << m << std::endl;
    std::cout << "Bведите количество условий отказа системы" << std::endl;
    std::cin >> kcm;
    std::cout << " Количество условий отказа системы " << kcm << std::endl;
    std::cout << "Введите количество установленных источников" << std::endl;
    std::cin >> ki;
    std::cout << " Количество установленных источников " << ki << std::endl;
    std::cout << "Введите потребляемую в режиме мощность" << std::endl;
    std::cin >> wd;
    std::cout << " Потребляемая в режиме мощность " << std::showpoint << std::fixed << std::setprecision(0) << wd << std::endl;
    std::cout << std::noshowpoint;
    std::cout << "Введите массив источников" << std::endl;
    std::string mgLine;
    std::getline(std::cin, mgLine);
    if(mgLine.empty()) std::getline(std::cin, mgLine);
    auto vals=parseNumbers(mgLine);
    for(size_t idx=0; idx<vals.size() && idx<static_cast<size_t>(ki); ++idx) mg[idx+1]=vals[idx];
    std::cout << " Массив источников:" << std::endl;
    for(int i=1;i<=ki;i++) {
        std::cout << " Мощность " << std::setw(2) << i << "-го источника " << std::showpoint << std::fixed << std::setprecision(0) << mg[i] << std::endl;
    }
    std::cout << std::noshowpoint;

    int jn=0, neli; double lambd, nui, pbi;
    while(true) {
        std::cout << "Введите номер элемента" << std::endl;
        std::cin >> neli;
        if(neli==0) break;
        jn++;
        nel[jn]=neli;
        std::cout << "Введите: лямбда, ну, Рв" << std::endl;
        { std::string line; std::getline(std::cin, line); if(line.empty()) std::getline(std::cin, line); auto arr=parseNumbers(line); if(arr.size()>0) lambd=arr[0]; if(arr.size()>1) nui=arr[1]; if(arr.size()>2) pbi=arr[2]; }
        lambda[neli]=lambd;
        nu[neli]=nui;
        pb[neli]=pbi;
        if(jn>=m) break;
    }

    std::cout << "Диагностическая матрица" << std::endl;
    std::cout << "Вввод Диагностической матрицы:" << std::endl;
    std::cout << "---------------" << std::endl;
    int nn1, nn2;
    for(int i=1;i<=kcm;i++) {
        std::cout << "Введите номера столбцов" << std::setw(2) << i << "-ой строки" << std::endl;
        { std::string line; std::getline(std::cin,line); if(line.empty()) std::getline(std::cin,line); auto arr=parseNumbers(line); nn1=arr.size()>0?static_cast<int>(arr[0]):0; nn2=arr.size()>1?static_cast<int>(arr[1]):0; }
        for(int j=1;j<=m;j++) mdi[j]=0;
        mdi[nn1]=1; mdi[nn2]=1;
        for(int j=1;j<=m;j++) {
            md[i][j]=mdi[j]==1;
        }
    }

    std::cout << "При определении источников" << std::endl;
    std::cout << "проверять состояние их щитов?" << std::endl;
    std::cout << "Да - введите 1" << std::endl;
    std::cout << "Нет - введите 0" << std::endl;
    int otv_input; std::cin >> otv_input;
    bool otv = otv_input!=0;

    std::cout << "-----------------------------------" << std::endl;
    std::cout << "|  №  |  лямбда |    ню   |  Рв   |" << std::endl;
    std::cout << "-----------------------------------" << std::endl;
    for(int j=1;j<=m;j++) {
        double amb=lambda[j]*1e6;
        double u1=nu[j]*1e3;
        std::cout << "| " << std::setw(3) << j << " | "
                  << std::setw(7) << std::fixed << std::setprecision(2) << amb << " | "
                  << std::setw(7) << std::fixed << std::setprecision(2) << u1 << " | "
                  << std::setw(5) << std::fixed << std::setprecision(3) << pb[j] << " |" << std::endl;
    }
    std::cout << "-----------------------------------" << std::endl;
    std::cout << "     РЕЗУЛЬТАТЫ МОДЕЛИРОВАНИЯ" << std::endl;
    std::cout << "---------------------------------------" << std::endl;
    std::cout << "| Д |  дельта | n исп. |  кг  |    В  |" << std::endl;
    std::cout << "---------------------------------------" << std::endl;

    std::mt19937 rng(42); // fixed seed for reproducibility
    std::uniform_real_distribution<double> dist(0.0,1.0);

    double ts=0.0, th=0.0, tm=0.0, ti=0.0, tsl=0.0;
    double sqt=0.0, l=0.0;
    int ns_int=100;

    while(true) {
        for(int n=1;n<=100;n++) {
            for(int i=1;i<=m;i++) {
                c[i]=true; t[i]=0.0;
            }
            th=0.0; tsl=0.0; bool cp=true; bool cq=true;
            for(int i=1;i<=m;i++) {
                double z;
                do { z = dist(rng); } while(z<=0.0);
                tau[i] = -std::log(z)/lambda[i];
                t[i] = tau[i];
            }
            while(true) {
                tm=t[1]; int k=1;
                for(int i=2;i<=m;i++) {
                    if(t[i]<tm) { tm=t[i]; k=i; }
                }
                if(tm>ta) {
                    if(!cp) {
                        ti=ta-th; ts+=ti; tsl+=ti;
                    }
                    break;
                }
                double z;
                if(!c[k]) {
                    c[k]=true;
                    do { z=dist(rng); } while(z<=0.0);
                    tau[k] = -std::log(z)/lambda[k];
                } else {
                    c[k]=false;
                    z = dist(rng);
                    if(z<pb[k]) {
                        do { z=dist(rng); } while(z<=0.0);
                        tau[k] = -std::log(z)/nu[k];
                    } else {
                        tau[k] = ta;
                    }
                }
                t[k]+=tau[k];
                bool ch;
                if(c[k] && cp) {
                    ch=true;
                } else {
                    double w=0.0;
                    if(otv) {
                        for(int i=1;i<=ki;i++) {
                            if(c[i] && c[ki+i]) w+=mg[i];
                        }
                    } else {
                        for(int i=1;i<=ki;i++) {
                            if(c[i]) w+=mg[i];
                        }
                    }
                    ch=true;
                    if(w<wd) ch=false;
                }
                if(!cp && ch) {
                    ti=tm-th; ts+=ti; tsl+=ti;
                } else if(cp && !ch) {
                    th=tm; cq=false;
                }
                cp=ch;
            }
            sqt += tsl*tsl;
            if(!cq) l += 1.0;
        }
        double kg = l - ts/ta/ns_int;
        double r = 1.0 - l/ns_int;
        double d = zdelta*std::sqrt(r*(l-r)/ns_int);
    std::cout << "| " << fortranF(d,7,5)
              << " | " << fortranF(delta,5,3)
              << " | " << fortranF(static_cast<double>(ns_int),6,0)
              << " | " << fortranF(kg,7,5)
              << " | " << fortranF(r,7,5) << " |" << std::endl;
        if(l>0.0 && d<=dd) break;
        ns_int += 100;
        if(ns_int>1000) break;
    }
    std::cout << "---------------------------------------" << std::endl;
    return 0;
}

